File: scripts/deploy.py
from brownie import accounts, config, network, RenToken, RenPool
from brownie.network.account import Account
from brownie.network.contract import Contract
import constants as C
import utils

active_network: str or None = network.show_active()
is_development: bool = active_network == C.NETWORKS["DEVELOPMENT"]

ren_token_addr = C.CONTRACT_ADDRESSES[active_network]["REN_TOKEN"]
darknode_registry_addr = C.CONTRACT_ADDRESSES[active_network]["DARKNODE_REGISTRY"]
claim_rewards_addr = C.CONTRACT_ADDRESSES[active_network]["CLAIM_REWARDS"]
gateway_addr = C.CONTRACT_ADDRESSES[active_network]["GATEWAY"]


def get_owner() -> Account:
    return (
        accounts[0] if is_development else accounts.add(config["wallets"]["from_key"])
    )


def get_node_operator() -> Account:
    return get_owner()


def get_ren_token(owner: Account) -> Contract or None:
    return (
        RenToken.deploy({"from": owner})
        if is_development
        else utils.load_contract(ren_token_addr)
    )


def main() -> tuple[Contract, Contract]:
    """
    Set your .env file accordingly before deploying the RenPool contract.
    In case of live networks, make sure your account is funded.
    """
    owner: Account = get_owner()
    node_operator: Account = get_node_operator()
    ren_token: Contract = get_ren_token(owner)

    ren_pool = RenPool.deploy(
        ren_token.address(),
        darknode_registry_addr,
        claim_rewards_addr,
        gateway_addr,
        owner,
        C.POOL_BOND,
        {"from": node_operator},
        publish_source = True,
    )

    return ren_token, ren_pool

File: scripts/deploy.js
const {
    ethers: {
        BigNumber: { from: bn }
    },
    network: {
        config: {
            accounts,
            renTokenAddr,
            darknodeRegistryAddr,
            darknodePaymentAddr,
            claimRewardsAddr,
            gatewayRegistryAddr,
        }
    } } = require('hardhat');
const chalk = require('chalk');

function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}

const DECIMALS = 18;
const DIGITS = bn(10).pow(DECIMALS);
const POOL_BOND = bn(100_000).mul(DIGITS);

async function main() {
    console.log(`${chalk.italic('\u{1F680} RenPool contract deployment')}`);
    console.log(`Using network ${chalk.bold(hre.network.name)} (${chalk.bold(hre.network.config.chainId)})`);

    console.log(`> Getting signers to deploy RenPool contract`);
    const owner = new ethers.Wallet(accounts[0], ethers.provider);
    const nodeOperator = owner;

    console.log(`> Deploying ${chalk.bold('RenPool')} contract`);
    const RenPool = await ethers.getContractFactory('RenPool');
    const renPool = await RenPool.connect(nodeOperator).deploy(
        renTokenAddr,
        darknodeRegistryAddr,
        darknodePaymentAddr,
        claimRewardsAddr,
        gatewayRegistryAddr,
        owner.address,
        POOL_BOND);
    await renPool.deployed();

    console.log(`> Deployed to ${chalk.bold(renPool.address)} TX ${chalk.bold(renPool.deployTransaction.hash)}`);

    if (hre.network.name === 'hardhat') {
        console.log('> Skipping RenPool contract Etherscan verification')
    } else {
        console.log('> Waiting before verification');
        await sleep(30000);
        const balance = await renPool.balanceOf(owner.address);
        console.log(`  Owner's balance is ${chalk.yellow(balance)}`);

        console.log('> Verifying RenPool smart contract in Etherscan')

        await hre.run("verify:verify", {
            address: renPool.address,
            constructorArguments: [
                renTokenAddr,
                darknodeRegistryAddr,
                darknodePaymentAddr,
                claimRewardsAddr,
                gatewayRegistryAddr,
                owner.address,
                POOL_BOND
            ],
        });
    }
}

main()
    .then(() => process.exit(0))
    .catch(err => {
        console.error(err);
        process.exit(1);
    });
